cloudify_azure/auth/oauth2.py

```diff
@@ -195,7 +200,116 @@ def request_access_token(self):
         return data
 
 
-class ServicePrincipalCertificateAuth(AADMixin):
+class CustomAADMixin(object):
+    VALID_CLOUD_ENVS = (
+        'AZURE_PUBLIC_CLOUD',
+        'AZURE_CHINA_CLOUD',
+        'AZURE_US_GOV_CLOUD',
+        'AZURE_GERMAN_CLOUD',
+    )
+
+    def _configure(self, **kwargs):
+        """Configure authentication endpoint.
+
+        Optional kwargs may include:
+
+            - cloud_environment (msrestazure.azure_cloud.Cloud):
+            A targeted cloud environment
+            - china (bool): Configure auth for China-based service,
+              default is 'False'.
+            - tenant (str): Alternative tenant, default is 'common'.
+            - resource (str): Alternative authentication resource, default
+              is 'https://management.core.windows.net/'.
+            - verify (bool): Verify secure connection, default is 'True'.
+            - timeout (int): Timeout of the request in seconds.
+            - proxies (dict): Dictionary mapping protocol or protocol and
+              hostname to the URL of the proxy.
+            - cache (adal.TokenCache): A adal.TokenCache,
+              see ADAL configuration
+              for details. This parameter is not used here and directly
+              passed to ADAL.
+        """
+        def _import_cloud_cls(name):
+            module = importlib.import_module('msrestazure.azure_cloud')
+            env_cls = getattr(module, name)
+            return env_cls
+
+        self._endpoints_active_directory = None
+        if kwargs.get('china'):
+            err_msg = ("china parameter is deprecated, "
+                       "please use "
+                       "cloud_environment"
+                       "=msrestazure.azure_cloud.AZURE_CHINA_CLOUD")
+            warnings.warn(err_msg, DeprecationWarning)
+            self._cloud_environment = AZURE_CHINA_CLOUD
+        else:
+            self._cloud_environment = AZURE_PUBLIC_CLOUD
+
+        # First check if the "cloud_environment" is specified by user so
+        # that we can use it instead of the default cloud environment
+        if kwargs.get('cloud_environment'):
+            cloud_env = kwargs['cloud_environment']
+            if cloud_env not in self.VALID_CLOUD_ENVS:
```

Comment:
What if `cloud_env` is `None` or some other falsy?

Reply:
If it is a `None` then it will not evaluate the whole condition but if it is `False` it will raise `NonRecoverableError`  Error